File: erlang/bob/bob.erl
-module(bob).
-export([response_for/1]).

% Yell -> "Whoa, chill out!"
% Question -> "Sure."
% blank -> "Fine. Be that way!"
% otherwise -> "Whatever."

response_for(String) -> case is_blank(String) of 
                          false -> case is_yell(String) of 
                               true -> "Whoa, chill out!";
                               false -> case is_question(String) of 
                                          true -> "Sure."; 
                                          false -> "Whatever." 
                                         end
                               end;
                          true -> "Fine. Be that way!"
                          end.

is_blank(String) -> string:strip(String) == "".

is_yell(String) -> string:equal(String, string:to_upper(String)) and has_at_least_one_letter(String).
has_at_least_one_letter(String) -> case re:run(String, "[a-zA-Z]+") of
                                     {match, _} -> true;
                                     nomatch -> false
                                  end.

is_question(String) -> lists:suffix("?", String).

File: csharp/bob/Bob.cs
using System;
using System.Linq;

public static class Bob
{
    private static string SILENCE_RESPONSE = "Fine. Be that way!";
    private static string Q_SHOUT_RESPONSE = "Calm down, I know what I'm doing!";
    private static string Q_RESPONSE = "Sure.";
    private static string SHOUT_RESPONSE = "Whoa, chill out!";
    private static string DEFAULT_RESPONSE = "Whatever.";

    public static string Response(string statement)
    {
        if (IsSilence(statement)) {
            return SILENCE_RESPONSE;
        }

        if (IsQuestion(statement) && IsShout(statement)) {
            return Q_SHOUT_RESPONSE;
        }

        if (IsQuestion(statement)) {
            return Q_RESPONSE;
        }

        if (IsShout(statement)) {
            return SHOUT_RESPONSE;
        }

        return DEFAULT_RESPONSE;
    }

    private static bool IsQuestion(string statement) {
        return statement.TrimEnd().EndsWith("?");
    }

    private static bool IsShout(string statement) {
        return statement.Any(c => char.IsLetter(c)) && !statement.Any(c => char.IsLower(c));
    }

    private static bool IsSilence(string statement) {
        return statement.Trim().Equals("");
    }
}